uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+  /// closure takes a `Span` pointing to the byte buffer, which it should fill
+  /// with bytes and return the actual number of bytes written.
+  ///
+  /// This function is (more than a little) convoluted. It's written this way
+  /// because we want to support UTF-8 and UTF-16 strings. The "size hint" is
+  /// the maximum number of bytes that the closure can write. For UTF-8 strings,
+  /// this is just the length. For UTF-16 strings, which must be converted to
+  /// UTF-8, this can be up to three times the length. Once the closure tells us
+  /// how many bytes it's actually written, we can write the length prefix, and
+  /// advance the current offset.
+  ///
+  /// Safety: The closure must copy the string's contents into the span, and
+  /// return the exact number of bytes it copied. Returning the wrong count can
+  /// either truncate the string, or leave uninitialized memory in the buffer.
+  /// The closure must not call any other methods on the writer.
+  void WriteRawString(size_t aSizeHint,
+                      const std::function<size_t(Span<char>)>& aClosure) {
+    // First, make sure the buffer is big enough to hold the length prefix.
+    // We'll start writing our string directly after the prefix.
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(uint32_t);
+    AssertInBounds(newOffset);
+    char* begin =
+        reinterpret_cast<char*>(&mBuffer.Elements()[newOffset.value()]);
+
+    // Next, ensure the buffer has space for enough bytes up to the size hint.
+    // We may write fewer bytes than hinted, but we need to handle the worst
+    // case if needed.
+    newOffset += aSizeHint;
+    AssertInBounds(newOffset);
+
+    // Call the closure to write the bytes directly into the buffer.
+    size_t bytesWritten = aClosure(Span(begin, aSizeHint));
+
+    // Great, now we know the real length! Write it at the beginning.
+    uint32_t lengthPrefix = PR_htonl(bytesWritten);
+    memcpy(&mBuffer.Elements()[mOffset.value()], &lengthPrefix,
+           sizeof(uint32_t));
+
+    // And figure out our actual offset.
+    newOffset -= aSizeHint;
+    newOffset += bytesWritten;
+    AssertInBounds(newOffset);
+    mOffset = newOffset;
+  }
+
+  RustBuffer ToRustBuffer() {
+    auto size = static_cast<uint32_t>(mOffset.value());
+    auto buffer = {{ ci.ffi_bytebuffer_alloc().name() }}(size);
+    memcpy(buffer.mData, mBuffer.Elements(), size);
+    return buffer;
+  }
+
+ private:
+  void AssertInBounds(const CheckedInt<size_t>& aNewOffset) const {
+    MOZ_RELEASE_ASSERT(aNewOffset.isValid() &&
+                       aNewOffset.value() <= mBuffer.Capacity());
+  }
+
+  template <typename T>
+  void WriteAt(const T& aValue,
+               const std::function<void(size_t, const T&)>& aClosure) {
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    mBuffer.SetLength(newOffset.value());
```

Comment:
Add a comment: we’ve reserved the capacity in `mBuffer` to hold the entire object, so this won’t allocate, but we still need to update the length or we’ll get an out-of-bounds exception indexing into the array.